llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp

```diff
@@ -6004,8 +6026,122 @@ struct VarArgSystemZHelper : public VarArgHelperBase {
   }
 };
 
-/// MIPS-specific implementation of VarArgHelper.
-/// NOTE: This is also used for LoongArch64.
+/// i386-specific implementation of VarArgHelper.
+struct VarArgI386Helper : public VarArgHelperBase {
```

Comment:
which one did you use as a source?

Reply:
If I remember correctly, this has been copied from GenericHelper. Then `visitCallBase` has been modified to differentiate "by val" parameter passing.